mantatail.py

```diff
@@ -2,92 +2,93 @@
-                        else:
-                            user_socket.sendall(command.error_not_registered())
+def recv_loop(state: ServerState, user_host: str, user_socket: socket.socket) -> None:
+    _user_message = None
+    _nick = None
 
-                        if _user_message and _nick:
-                            user = User(user_host, user_socket, _user_message, _nick)
-                            command.motd(self, user)
+    user = None
 
+    with user_socket:
+        while True:
+            request = b""
+            # IRC messages always end with b"\r\n" (netcat uses "\n")
+            while not request.endswith(b"\n"):
+                request_chunk = user_socket.recv(4096)
+                if request_chunk:
+                    request += request_chunk
+                else:
+                    if user is not None:
+                        print(f"{user.nick} has disconnected.")
+                    else:
+                        print("Disconnected.")
+                    return
+
+            decoded_message = request.decode("utf-8")
+            for line in split_on_new_line(decoded_message)[:-1]:
+                if " " in line:
+                    (verb, message) = line.split(" ", 1)
+                else:
+                    verb = line
+                    message = verb
+
+                verb_lower = verb.lower()
+                parsed_command = "handle_" + verb_lower
+
+                if user is None:
+                    if verb_lower == "user":
+                        _user_message = message
+                    elif verb_lower == "nick":
+                        _nick = message
                     else:
-                        try:
-                            # ex. "command.handle_nick" or "command.handle_join"
-                            call_handler_function = getattr(command, parsed_command)
-                        except AttributeError:
-                            command.error_unknown_command(user, verb_lower)
-                        else:
-                            call_handler_function(self, user, message)
+                        user_socket.sendall(command.error_not_registered())
+
+                    if _user_message and _nick:
+                        user = User(user_host, user_socket, _user_message, _nick)
+                        command.motd(state.motd_content, user)
+
+                else:
+                    try:
+                        # ex. "command.handle_nick" or "command.handle_join"
+                        call_handler_function = getattr(command, parsed_command)
+                    except AttributeError:
+                        command.error_unknown_command(user, verb_lower)
+                    else:
+                        call_handler_function(state, user, message)
 
-                        if user.closed_connection:
-                            return
+                    if user.closed_connection:
+                        return
 
 
 class User:
```

Comment:
```suggestion
class UserConnection:
```

I think it makes sense to rename this in this refactor to clarify its responsibility